WordPress/gutenberg
# ToC block: re-enable + use static markup and only support core Heading and Page Break blocks

## Description
Fixes #29560 and re-enables the Table of Contents block, which was first merged over a year ago in #21234, only to be temporarily (heh) disabled 2 weeks later in #29718 to provide more time for some implementation issues to be resolved.

This time, the block is using static markup and only supports the core Heading and Page Break blocks, using a much cleaner and performant implementation. I've also switched to using `<ol>` for the lists.

## Implementation notes

### Why `<ol>`s?

See https://github.com/WordPress/gutenberg/pull/29739#issuecomment-1067347781 and [this other comment](https://github.com/WordPress/gutenberg/pull/29739#issuecomment-1083394967) for additional notes on that decision.

### Why absolute links?

See https://github.com/WordPress/gutenberg/pull/29739#issuecomment-1083463587.

### Undo behavior quirk

I have noticed one undesirable quirk with the current implementation, though I'm not sure if there's any way around it: when a Table of Contents block is present, every letter typed in a Heading block counts as a separate undo step, in contrast to the default behavior of multiple letters typed in sequence being combined into a single undo step. Is this a blocker? If so, I'd appreciate any suggestions on how to fix it, because I'm not aware of any solutions here. (Remember, it only affects Headings, not other blocks.)

I've opened #41031 to track this bug.

### Static rendering

This PR currently uses static rendering for the Table of Contents, meaning the markup is only altered in the editor, and on save, the final markup is saved to the post content.

#### Advantages

- Performance. The heading list and markup are calculated only in the editor, meaning the Table of Contents requires no more effort to render on the front-end than a Paragraph or Heading.
- Fallback markup. Even if the blocks of a post fail to be parsed properly, the block will still render its static content between the block delimiter comments (again, just like a Paragraph).
- Simplicity. It's really nice to do everything in JS and not have to re-implement the block in PHP as well.


#### Disadvantages

- If we ever want to use the Table of Contents in the context of something like a post template (outside of the Post Content block), static markup won't work.
- Though links breaking when you change the permalink structure is "normal" for WordPress, that doesn't mean we shouldn't try to avoid this when possible. It would be really nice if the Table of Contents block just handled this automatically - and the only way to do this is via dynamic rendering.
- Dynamic rendering would allow us to create a Table of Contents block that can list the headings of any post, not just the current one. This would be very fragile with static rendering, since updating the referenced post would not update the Table of Contents.

## Testing instructions

Try adding Heading and Page Break blocks to create a paginated post. Set various heading levels on the Headings. Set anchors/ids on some of the Headings. Place a Table of Contents block somewhere between the Page Breaks. Ensure that the links to the headings work on the front-end, regardless of which page the headings are on. (Note that only headings with anchors will be linked.) Try toggling the "Only include current page" option to make the block only list headings on the same page as itself.

Note that in a recent commit, I've added a feature where inserting a Table of Contents block turns on the Heading block's automatic anchor generation - a feature which is already in core, but is disabled by default, with no visible UI control available to turn it on permanently yet.

## Screenshot

![image](https://user-images.githubusercontent.com/19592990/159329411-060a36d0-dcac-40a4-b9ef-960225607523.png)

## Checklist:
- [n/a] My code follows the accessibility standards. 
- [n/a] I've tested my changes with keyboard and screen readers.
